File: src/main/java/com/portofino/realtrainmodunofficial/model/MQOModel.java
package com.portofino.realtrainmodunofficial.model;

import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.blaze3d.vertex.VertexConsumer;
import net.minecraft.resources.ResourceLocation;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class MQOModel {
    private final List<Vertex> vertices = new ArrayList<>();
    private final List<Face> faces = new ArrayList<>();
    private final Map<String, Material> materials = new HashMap<>();
    private final List<ObjectGroup> groups = new ArrayList<>();
    private Object scriptEngine; // Script engine for animations

    public void addVertex(float x, float y, float z) {
        vertices.add(new Vertex(x, y, z));
    }

    public void addFace(int[] vertexIndices, float[][] uvs, String materialName) {
        faces.add(new Face(vertexIndices, uvs, materialName));
    }

    public void addMaterial(String name, ResourceLocation texture) {
        materials.put(name, new Material(name, texture));
    }

    public void addGroup(String name) {
        groups.add(new ObjectGroup(name));
    }

    public void render(PoseStack poseStack, VertexConsumer consumer, int packedLight, int pass) {
        // Execute script before rendering if available
        if (scriptEngine != null) {
            executeScript(poseStack, pass);
        }

        com.portofino.realtrainmodunofficial.RealTrainModUnofficial.LOGGER.info("Rendering MQO model: {} vertices, {} faces, {} materials", vertices.size(), faces.size(), materials.size());

        for (Face face : faces) {
            Material mat = materials.get(face.materialName);
            if (mat != null) {
                // Render face with texture
                for (int i = 0; i < face.vertexIndices.length; i++) {
                    int vi = face.vertexIndices[i];
                    if (vi >= 0 && vi < vertices.size()) {
                        Vertex v = vertices.get(vi);
                        float u = face.uvs != null && i < face.uvs.length ? face.uvs[i][0] : 0.0f;
                        float vCoord = face.uvs != null && i < face.uvs.length ? face.uvs[i][1] : 0.0f;
                        consumer.addVertex(poseStack.last().pose(), v.x, v.y, v.z)
                            .setColor(1.0f, 1.0f, 1.0f, 1.0f)
                            .setUv(u, vCoord)
                            .setUv2((packedLight & 0xFFFF), (packedLight >> 16))
                            .setNormal(0.0f, 1.0f, 0.0f);
                    }
                }
            }
        }
    }

    private void executeScript(PoseStack poseStack, int pass) {
        // Execute script to apply transformations
        // This is a simplified implementation
        try {
            if (scriptEngine instanceof javax.script.ScriptEngine) {
                javax.script.ScriptEngine engine = (javax.script.ScriptEngine) scriptEngine;
                engine.put("poseStack", poseStack);
                engine.put("pass", pass);
                engine.eval("if (typeof render === 'function') render(poseStack, pass);");
            }
        } catch (Exception e) {
            // Script execution failed, continue without script
        }
    }

    public void setScriptEngine(Object engine) {
        this.scriptEngine = engine;
    }

    public void render(PoseStack poseStack, VertexConsumer consumer, int packedLight) {
        render(poseStack, consumer, packedLight, 0);
    }

    public static class Vertex {
        public final float x, y, z;
        public Vertex(float x, float y, float z) {
            this.x = x;
            this.y = y;
            this.z = z;
        }
    }

    public static class Face {
        public final int[] vertexIndices;
        public final float[][] uvs; // UV coordinates for each vertex
        public final String materialName;
        public Face(int[] vertexIndices, float[][] uvs, String materialName) {
            this.vertexIndices = vertexIndices;
            this.uvs = uvs;
            this.materialName = materialName;
        }
    }

    public static class Material {
        public final String name;
        public final ResourceLocation texture;
        public Material(String name, ResourceLocation texture) {
            this.name = name;
            this.texture = texture;
        }
    }

    public static class ObjectGroup {
        public final String name;
        public ObjectGroup(String name) {
            this.name = name;
        }
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/model/MQOModel.kt
package cc.mirukuneko.realtrainmodrenewed.model

import com.mojang.blaze3d.vertex.PoseStack
import com.mojang.blaze3d.vertex.VertexConsumer
import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed
import net.minecraft.resources.Identifier

class MQOModel {
    data class Vertex(@JvmField val x: Float, @JvmField val y: Float, @JvmField val z: Float)

    data class Face(
        @JvmField val vertexIndices: IntArray,
        @JvmField val uvs: Array<FloatArray>?,
        @JvmField val materialName: String
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is Face) return false
            return vertexIndices.contentEquals(other.vertexIndices) &&
                uvs.contentDeepEquals(other.uvs) &&
                materialName == other.materialName
        }

        override fun hashCode(): Int {
            var result = vertexIndices.contentHashCode()
            result = 31 * result + (uvs?.contentDeepHashCode() ?: 0)
            result = 31 * result + materialName.hashCode()
            return result
        }
    }

    data class Material(@JvmField val name: String, @JvmField val texture: Identifier)

    data class ObjectGroup(@JvmField val name: String)

    private val vertices: MutableList<Vertex> = ArrayList()
    private val faces: MutableList<Face> = ArrayList()
    private val materials: MutableMap<String, Material> = HashMap()
    private val groups: MutableList<ObjectGroup> = ArrayList()
    private var scriptEngine: Any? = null

    fun addVertex(x: Float, y: Float, z: Float) {
        vertices.add(Vertex(x, y, z))
    }

    fun addFace(vertexIndices: IntArray, uvs: Array<FloatArray>?, materialName: String) {
        faces.add(Face(vertexIndices, uvs, materialName))
    }

    fun addMaterial(name: String, texture: Identifier) {
        materials[name] = Material(name, texture)
    }

    fun addGroup(name: String) {
        groups.add(ObjectGroup(name))
    }

    @JvmOverloads
    fun render(poseStack: PoseStack, consumer: VertexConsumer, packedLight: Int, pass: Int = 0) {
        if (scriptEngine != null) executeScript(poseStack, pass)
        RealTrainModRenewed.LOGGER.debug(
            "Rendering MQO model: {} vertices, {} faces, {} materials",
            vertices.size, faces.size, materials.size
        )
        for (face in faces) {
            val mat = materials[face.materialName] ?: continue
            for (i in face.vertexIndices.indices) {
                val vi = face.vertexIndices[i]
                if (vi < 0 || vi >= vertices.size) continue
                val v = vertices[vi]
                val u = face.uvs?.let { if (i < it.size) it[i][0] else 0.0f } ?: 0.0f
                val vCoord = face.uvs?.let { if (i < it.size) it[i][1] else 0.0f } ?: 0.0f
                consumer.addVertex(poseStack.last().pose(), v.x, v.y, v.z)
                    .setColor(1.0f, 1.0f, 1.0f, 1.0f)
                    .setUv(u, vCoord)
                    .setUv2((packedLight and 0xFFFF), (packedLight shr 16))
                    .setNormal(0.0f, 1.0f, 0.0f)
            }
        }
    }

    private fun executeScript(poseStack: PoseStack, pass: Int) {
        try {
            val engine = scriptEngine
            if (engine is javax.script.ScriptEngine) {
                engine.put("poseStack", poseStack)
                engine.put("pass", pass)
                engine.eval("if (typeof render === 'function') render(poseStack, pass);")
            }
        } catch (_: Exception) {
            // Script execution failed, continue without script
        }
    }

    fun setScriptEngine(engine: Any?) {
        scriptEngine = engine
    }
}
